### ISSUE TITLE
BigDecimal: #coerce method is not called when BigDecimal is left hand operand

### ISSUE THREAD

**Issue by deepj** [oracle/truffleruby#1507]
See @janko-m's comment with an example to reproduce it https://github.com/oracle/truffleruby/issues/1507#issuecomment-445162208

**Old report**
I've tried TrufflyRuby `1.0.0-rc10` and Sequel's specs. Good news, the most specs pass successfully now. But with one exception: `BigDecimal('1.0') * Sequel[:a][:y]`

This happens only with `BigDecimal` until now. 

To reproduce

```ruby
require 'sequel'
BigDecimal('1.0') * Sequel[:a][:y]
```

Error:

```
NoMethodError: undefined method `to_f' for #<Sequel::SQL::QualifiedIdentifier @column=>:y, @table=>"a">:Sequel::SQL::QualifiedIdentifier
Did you mean?  to_s
	from (irb):2:in `*'
	from (irb):2
	from ~/.rubies/truffleruby-1.0.0-rc10/bin/irb:29:in `<main>'
```

There is no issue e.g. with `float`

```
1.0 * Sequel[:a][:y]
=> #<Sequel::SQL::NumericExpression @args=>[1.0, #<Sequel::SQL::QualifiedIdentifier @column=>:y, @table=>"a">], @op=>:*>
```

**Comment by janko-m** [comment]
The issue is that TruffleRuby doesn't call the `#coerce` method on the right hand operand when left hand operand is a `BigDecimal`, as this example shows:

```rb
require "bigdecimal"

class MyNumber
  def initialize(number)
    @number = number
  end

  def coerce(other)
    puts "called with #{other.inspect}"

    case other
    when Integer    then [other, @number]
    when BigDecimal then [other, BigDecimal.new(@number)]
    end
  end
end

1 + MyNumber.new(2)
BigDecimal.new(1) + MyNumber.new(2)
```
```
called with 1
bla.rb:19:in `+': undefined method `to_f' for #<MyNumber:0x88 @number=2>:MyNumber (NoMethodError)
Did you mean?  to_s
        from bla.rb:19:in `<main>'
```

**Comment by deepj** [comment]
@janko-m thanks for the example. I wanted to do something similar tonight.

**Comment by deepj** [comment]
Just for a note: after fixing this, Sequel’s specs would pass on TruffleRuby

**Comment by Quintasan** [comment]
I'd like to take a stab at fixing this but I have never worked on TruffleRuby (nor any other Ruby implementation) before. Any tips on where should I start?

**Comment by chrisseaton** [comment]
@Quintasan you can join us in gitter.im/graalvm/truffleruby for interactive help in building and working on TruffleRuby.

This issue is probably easy to fix, but may require going down a bit of a rabbit hole of some of the strange recursive method calls and primitives involved in coercion.

**Comment by deepj** [comment]
Just for a note, there are other issues with BigDecimal in `bson` gem. Not reported by me yet...

**Comment by eregon** [comment]
Reopening, as only + and - were fixed, not yet other operators. Also see #1598.

**Comment by eregon** [comment]
Fully fixed in #1598.
